### PR TITLE
feat(bathymetry): create a process to convert gebco into hillshaded rasters

### PR DESCRIPTION
**Pull request by blacha**

### CODE REVIEW

File: packages/bathymetry/src/bathy.maker.ts [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        return this.filePath + '.tiff';
+    }
+
+    /** File name of the path */
+    get fileName(): string {
+        return path.basename(this.filePath);
+    }
+
+    async render(logger: LogType): Promise<void> {
+        console.log(this.filePath);
```
Comment (jacott) [inline packages/bathymetry/src/bathy.maker.ts:53]: Bare console.log

File: packages/bathymetry/src/mapnik.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    <Datasource>
+      <Parameter name="file"><![CDATA[${sourceFile}]]></Parameter>
+      <Parameter name="band"><![CDATA[1]]></Parameter>
+      <Parameter name="type"><![CDATA[gdal]]></Parameter>
+    </Datasource>
+  </Layer>
+</Map>`;
+}
+
+/** compoiste a hillshade and base tile into a hillshaded file with mapnik */
```
Comment (jacott) [inline packages/bathymetry/src/mapnik.ts:64]: composite miss-spelt

File: packages/bathymetry/src/mapnik.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+async function render(bm: BathyMaker, tile: Tile, logger: LogType): Promise<string> {
+    const tileId = TileMatrixSet.tileToName(tile);
+    const warpedPath = bm.file.name(FileType.Warped, tileId);
+    const hillShadePath = bm.file.name(FileType.HillShade, tileId);
+    const outputPath = bm.file.name(FileType.Rendered, tileId);
+    if (fs.existsSync(outputPath)) return outputPath;
+
```
Comment (jacott) [inline packages/bathymetry/src/mapnik.ts:71]: Do we want to continue if outputPath exists?

File: packages/bathymetry/src/bathy.maker.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            }
+        }
+        await Promise.all(promises).catch((err) => logger.fatal(err, 'FailedToRun'));
+    }
+
+    /** Create a multi hash of the source file  */
+    async createSourceHash(logger: LogType): Promise<string> {
+        const hashPath = this.file.name(FileType.Hash);
+        if (fs.existsSync(hashPath)) return (await fs.promises.readFile(hashPath)).toString();
+        logger.info({ hashPath }, 'CreateHash');
```
Comment (jacott) [inline packages/bathymetry/src/bathy.maker.ts:89]: Do we want to not re-create the hash if it exists already; maybe it changed?

File: packages/bathymetry/src/bathy.maker.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            await this.createCompositeTile(tile, logger);
+            await this.createTileMetadata(tile, logger);
+        } catch (err) {
+            logger.error({ err }, 'Failed');
+        }
+    }
+
+    /** convert a input file into one that can be processed by this script */
+    async createSourceGeoTiff(logger: LogType): Promise<void> {
+        if (fs.existsSync(this.tiffPath)) return;
```
Comment (jacott) [inline packages/bathymetry/src/bathy.maker.ts:110]: Do we want to continue if path exists?

File: packages/bathymetry/src/bathy.maker.ts [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            ],
+            logger,
+        );
+    }
+
+    /** Create a tile in the output TMS's CRS */
+    async createTile(tile: Tile, logger: LogType): Promise<void> {
+        const tileId = TileMatrixSet.tileToName(tile);
+        const warpedPath = this.file.name(FileType.Warped, tileId);
+        if (fs.existsSync(warpedPath)) return;
```
Comment (jacott) [inline packages/bathymetry/src/bathy.maker.ts:133]: Do we want to continue if path exists?

File: packages/bathymetry/src/bathy.maker.ts [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+        await gdal.run('gdalwarp', warpCommand, logger);
+    }
+
+    /** Create a hillshade for a tile */
+    async createHillShadedTile(tile: Tile, logger: LogType): Promise<void> {
+        const tileId = TileMatrixSet.tileToName(tile);
+        const warped = this.file.name(FileType.Warped, tileId);
+        const target = this.file.name(FileType.HillShade, tileId);
+
+        if (fs.existsSync(target)) return;
```
Comment (jacott) [inline packages/bathymetry/src/bathy.maker.ts:166]: Do we want to continue if path exists?

File: packages/bathymetry/src/bathy.maker.ts [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        await gdal.run('gdaldem', ['hillshade', '-compute_edges', '-multidirectional', warped, target], logger);
+    }
+
+    /** Use mapnik to composite the hillshaded tile and the rendered tile */
+    async createCompositeTile(tile: Tile, logger: LogType): Promise<void> {
+        const tileId = TileMatrixSet.tileToName(tile);
+
+        const renderedPath = await MapnikRender.render(this, tile, logger);
+        const outputPath = this.file.name(FileType.Output, tileId);
+        if (fs.existsSync(outputPath)) return;
```
Comment (jacott) [inline packages/bathymetry/src/bathy.maker.ts:181]: Do we want to continue if path exists?

File: packages/bathymetry/src/mapnik.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    await new Promise((resolve, reject) =>
+        map.fromString(template, (err: Error) => (err == null ? resolve() : reject(err))),
+    );
+
+    map.zoomAll();
+
+    const startTime = Date.now();
+    await new Promise((resolve, reject) =>
+        map.renderFile(outputPath, (err: Error) => (err == null ? resolve() : reject(err))),
+    );
```
Comment (jacott) [inline packages/bathymetry/src/mapnik.ts:84]: You could use `util.promisify` but I am ambivalent about  it

File: packages/bathymetry/src/stac.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+/** Write some basic metadata about how the file was created*/
+async function create(bm: BathyMaker, tile: Tile, logger: LogType): Promise<Record<string, unknown>> {
+    const tileId = TileMatrixSet.tileToName(tile);
+    const outputTiffPath = bm.file.name(FileType.Output, tileId);
+    const ptms = ProjectionTileMatrixSet.get(bm.tms.projection.code);
+
+    const bounds = ptms.tileToWgs84Bounds(tile);
+    const geometry = bounds.toPolygon();
```
Comment (jacott) [inline packages/bathymetry/src/stac.ts:23]: I think this needs a bit more `geojson` around it; guessing: `{type: 'Polygon', coordinates: bounds.toPolygon()[0]}`

File: packages/shared/src/proj/projection.tile.matrix.set.ts [hunk 11]
```diff
@@ -73,6 +73,20 @@ export class ProjectionTileMatrixSet {
...
+        const [lrLon, lrLat] = this.proj.toWsg84([lr.x, lr.y]);
+        return new Bounds(
+            Math.min(ulLon, lrLon),
+            Math.min(ulLat, lrLat),
+            Math.abs(lrLon - ulLon),
+            Math.abs(lrLat - ulLat),
+        );
+    }
+
     /**
```
Comment (jacott) [inline packages/shared/src/proj/projection.tile.matrix.set.ts:90]: Missing tests

File: packages/bathymetry/src/bathy.maker.ts (hunk 1 above)
Comment (blacha) [reply packages/bathymetry/src/bathy.maker.ts]: cool, removed

File: packages/bathymetry/src/bathy.maker.ts (hunk 4 above)
Comment (blacha) [reply packages/bathymetry/src/bathy.maker.ts:102]: I was thinking about this, I could check a fs stat to see modified time or something, I think what ideally i want to do is delete the `.bathy` folder after a success

File: packages/bathymetry/src/bathy.maker.ts (hunk 7 above)
Comment (blacha) [reply packages/bathymetry/src/bathy.maker.ts:181]: this is a lazy processing thingy, if the file already exists skip the reprocessing of it, quite useful for testing it.

**Pull request merged**